### DISCUSSION TITLE
Applying styles to `<body>`

### DISCUSSION THREAD

**Discussion by sharath-jonnala07** [facebook/stylex#399] [Q&A]
I am unable to apply margin 0 to the entire body like in CSS. How to do that?

**Comment by nmn** [comment]
What is your project setup? In something like Next, you can apply styles on body like any other component.

If your `body` element is not implemented within a component, you should use a `CSS` file with traditional CSS resets.

**Comment by sharath-jonnala07** [comment]
I am using react

**Reply by nmn** [reply]
That is not enough information.

**Comment by addictionai** [comment]
Hey @Sharathjonnala2003. 

In case you still need help, you can accomplish this by adding a @layer directive in your root css file. As @nmn mentioned, it would serve as a css 'reset'.

Here's a practical implementation that would work in any React app framework.

1. Locate the top level css import for your project
  - Vite: /src/index.css, imported on /src/main.tsx
  - Remix: /app/index.css, imported on /app/root.tsx
2. Add your base layer to index.css  (the order of @layer matters)
3. Done!

```css

@stylex stylesheet;

@layer base {
 body {
    margin: 0,
    padding: 0,
  }
}
```

What this does, is it adds a [CSS Layer](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer) which contains rules you want in the order you want them to take place. By having @layer base first, this will apply to all styles from that point forward. Other common layers may be:

```css
@layer theme {
   /* your app or brand theme */
}

@layer utils {
   /* common utility classes you want to reuse */
}
```

Hope this helps! Happy coding :)

**Reply by r11234** [reply]
I don't have much experience with styleX so maybe I am missing something, but is the use of layers required here? I am able to apply styles to `body` element without adding those styles to a layer.

**Reply by nmn** [reply]
@r11234 CSS layers are not required. You can configure the various bundler plugins to generate layers or not. If you opt-out of CSS Layers, we use CSS selector specificity bumping instead, which is a how lightningcss/postcss would polyfill layers.

This means that in either case, you have to be a bit careful about the specificity of your styles.

- Unlayered styles *always* override layered styles
- You styles would normally lose to StyleX styles if you configure it to not use Layers.

**Comment by nmn** [comment]
@addictionai The one thing to be careful about is that CSS Layers are order dependent so ensure that the StyleX styles are injected after your base styles.

**Comment by predaytor** [comment]
Is it possible to set `::selection` in StyleX? This example didn't work:
```ts
const styles = stylex.create({
	html: {		
		'::selection': {
			color: 'pink',
			backgroundColor: 'black',
		}
	},
});

```

**Reply by nmn** [reply]
This should work. Got more details or a repro for me? A minimal repro with no other styles might help see what CSS is generated.

Was `styles.html` correctly applied on the `<html>` element?

**Update**: I did a quick test to see what was happening and I see the correct CSS is generated and applied but the styling isn't taking effect. This might be an unrelated quirk of CSS, but let me see how to do this right.

**Update 2**: Various HTML elements like `<p>` have browser styles for setting `::selection` so setting `::selection` on the `<body>` element doesn't work. Using `html::selection` doesn't seem to work at all.

This is probably why `::selection` is usually used without any other selector which makes it work universally. However, this is a "global" style and so should be defined in a CSS file outside of StyleX.

If, however, you're trying to override `::selection` for a particular element like `<div>` or `<p>` then using it through StyleX should work as expected.

**Reply by predaytor** [reply]
Got it. Thanks for answer.

**Reply by TxHawks** [reply]
This means that it is impossible to use a color variable defined by StyleX, which isn't ideal from a design system point of view

**Reply by TxHawks** [reply]
One way around this is to assign a custom property to `::selection` in some global css file, say something like:
```css
::selection { 
  background-color: --selectionBg; 
 color: --selectionFg;
}
```

And then set the value of these variables on the `html` element using StyleX, preferably with a token created with `defineVars`. That works perfectly and also allows for easy local overrides.

**Reply by nmn** [reply]
We're looking into ways to allow usage of variables defined using `stylex.defineVars` to be usable within CSS files. We don't have a good solution for that yet.
